src/plugins/multifile/multifile.c

```diff
@@ -0,0 +1,693 @@
+/**
+ * @file
+ *
+ * @brief Source for multifile plugin
+ *
+ * @copyright BSD License (see doc/LICENSE.md or http://www.libelektra.org)
+ *
+ */
+
+#include "multifile.h"
+
+#include <dirent.h>
+#include <fnmatch.h>
+#include <kdbconfig.h>
+#include <kdbhelper.h>
+#include <kdbinternal.h>
+#include <kdbmodule.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#define DEFAULT_RESOLVER "resolver"
+#define DEFAULT_PATTERN "*"
+#define DEFAULT_STORAGE "ini"
+
+typedef enum {
+	SETRESOLVER = 0,
+	SETSTORAGE,
+	COMMIT,
+} SetPhases;
+
+typedef enum {
+	GETRESOLVER = 0,
+	GETSTORAGE,
+} GetPhases;
+
+typedef enum {
+	EMPTY = -2,
+	ERROR = -1,
+	NOUPDATE = 0,
+	SUCCESS = 1,
+} Codes;
+
+
+typedef struct
+{
+	char * directory;
+	char * pattern;
+	SetPhases setPhase;
+	GetPhases getPhase;
+	KeySet * modules;
+	KeySet * childBackends;
+	char * resolver;
+	char * storage;
+	unsigned short stayAlive;
+	unsigned short hasDeleted;
+} MultiConfig;
+
+typedef struct
+{
+	char * filename;
+	char * fullPath;
+	char * parentString;
+	char * tmpFilename;
+	Plugin * resolver;
+	Codes rcResolver;
+	Plugin * storage;
+	Codes rcStorage;
+	KeySet * ks;
+} SingleConfig;
+
+static inline Codes rvToRc (int rc)
+{
+	switch (rc)
+	{
+	case -2:
+		return EMPTY;
+		break;
+	case -1:
+		return ERROR;
+		break;
+	case 0:
+		return NOUPDATE;
+		break;
+	case 1:
+		return SUCCESS;
+		break;
+	}
+	return ERROR;
+}
+
+int elektraMultifileCheckFile (const char * filename)
+{
+	if (!filename) return -1;
+	if (filename[0] == '/') return 0;
+	return 1;
+}
+
+int elektraMultifileOpen (Plugin * handle ELEKTRA_UNUSED, Key * errorKey ELEKTRA_UNUSED)
+{
+	// plugin initialization logic
+	// this function is optional
+
+	return 1; // success
+}
+
+static void closeBackend (SingleConfig * s)
+{
+	if (!s) return;
+	if (s->fullPath) elektraFree (s->fullPath);
+	if (s->filename) elektraFree (s->filename);
+	if (s->parentString) elektraFree (s->parentString);
+	if (s->tmpFilename) elektraFree (s->tmpFilename);
+	if (s->resolver) elektraPluginClose (s->resolver, NULL);
```

Comment:
most of the other plugins use elektraPluginClose (<plugin>, 0) instead of NULL, not sure whats the right thing to do. And by the way, are all those null checks necessary? I think for plugins they might be fine, if plugin developers have no checks, but i guess elektraFree simply does nothing if its already null?